python/cudf/cudf/core/scalar.py

```diff
@@ -0,0 +1,344 @@
+        return f"Scalar({str(self.value)}, dtype={self.dtype})"
+
+    def _binop_result_dtype_or_error(self, other, op):
+        if op in {"__eq__", "__ne__", "__lt__", "__gt__", "__le__", "__ge__"}:
+            return np.bool
+
+        out_dtype = get_allowed_combinations_for_operator(
+            self.dtype, other.dtype, op
+        )
+
+        # datetime handling
+        if out_dtype in {"M", "m"}:
+            if self.dtype.char in {"M", "m"} and other.dtype.char not in {
+                "M",
+                "m",
+            }:
+                return self.dtype
+            if other.dtype.char in {"M", "m"} and self.dtype.char not in {
+                "M",
+                "m",
+            }:
+                return other.dtype
+            else:
+                if (
+                    op == "__sub__"
+                    and self.dtype.char == other.dtype.char == "M"
+                ):
+                    res, _ = np.datetime_data(max(self.dtype, other.dtype))
+                    return np.dtype("m8" + f"[{res}]")
+                return np.result_type(self.dtype, other.dtype)
+
+        return np.dtype(out_dtype)
+
+    def _scalar_binop(self, other, op):
+        if isinstance(other, (ColumnBase, Series, Index, np.ndarray)):
+            # dispatch to column implementation
+            return NotImplemented
+        other = to_cudf_compatible_scalar(other)
+        out_dtype = self._binop_result_dtype_or_error(other, op)
+        valid = self.is_valid and (
+            isinstance(other, np.generic) or other.is_valid
+        )
+        if not valid:
+            return Scalar(None, dtype=out_dtype)
+        else:
+            result = self._dispatch_scalar_binop(other, op)
+            return Scalar(result, dtype=out_dtype)
+
+    def _dispatch_scalar_binop(self, other, op):
+        if isinstance(other, Scalar):
+            other = other.value
+        return getattr(self.value, op)(other)
+
+    def _unaop_result_type_or_error(self, op):
+        if op == "__neg__" and self.dtype == "bool":
+            raise TypeError(
+                "Boolean scalars in cuDF do not support"
+                " negation, use logical not"
+            )
+
+        if op in {"__ceil__", "__floor__"}:
+            if self.dtype.char in "bBhHf?":
+                return np.dtype("float32")
+            else:
+                return np.dtype("float64")
+        return self.dtype
+
+    def _scalar_unaop(self, op):
+        out_dtype = self._unaop_result_type_or_error(op)
+        if not self.is_valid():
+            result = None
+        else:
+            result = self._dispatch_scalar_unaop(op)
+            return Scalar(result, dtype=out_dtype)
+
+    def _dispatch_scalar_unaop(self, op):
+        if op == "__floor__":
+            return np.floor(self.value)
+        if op == "__ceil__":
+            return np.ceil(self.value)
+        return getattr(self.value, op)()
+
+
+class cudf_NA_type(object):
```

Comment:
Suggest not prefixing with `cudf`; also using CamelCase as recommended by PEP8.

```suggestion
class _NAType(object)
```

Also needs change below where we define `NA`.